### PR TITLE
Make AppearsOnRadar Conditional & Player Stance Awear

### PR DESCRIPTION
**Pull request by atlimit8**
This can be used to implement radar invisibility as mentioned in #13209.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs [hunk 1]
```diff
@@ -14,28 +14,79 @@
...
-	public class AppearsOnRadar : IRadarSignature, INotifyCreated
+	public class AppearsOnRadar : IObservesVariables, IRadarSignature, INotifyCreated
 	{
+		static readonly IEnumerable<Pair<CPos, Color>> NoCells = Enumerable.Empty<Pair<CPos, Color>>();
 		readonly AppearsOnRadarInfo info;
 		IRadarColorModifier modifier;
 
 		Color currentColor = Color.Transparent;
 		Func<Pair<CPos, SubCell>, Pair<CPos, Color>> cellToSignature;
+		Stance visibleStances = Stance.Ally | Stance.Enemy;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs]: What about Neutral?

**Review by pchote — COMMENTED** [review]
What if we added a fixed `ValidStances` field and made the trait Conditional, and then have separate trait instances for enemy/neutral vs ally for actors that can become radar invisible?

File: OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs (hunk 1 above)
Comment (atlimit8) [inline OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs]: Unless I am missing something, human players are either allies or enemies.

**Comment by atlimit8** [comment]
I am OK with the idea of using `ValidStances` and basing it on `ConditionalTrait`.

File: OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs]: There are quite a few maps that place neutral base structures or units on the map for decoration or to be captured.

Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs]: Someone could (in theory) force set stances in map.yaml.

**Comment by atlimit8** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs [hunk 2]
```diff
@@ -18,38 +18,42 @@
...
-		public void Created(Actor self)
+		protected override void Created(Actor self)
 		{
+			base.Created(self);
 			modifier = self.TraitsImplementing<IRadarColorModifier>().FirstOrDefault();
 		}
 
 		public IEnumerable<Pair<CPos, Color>> RadarSignatureCells(Actor self)
 		{
+			if (IsTraitDisabled || !Info.ValidStances.HasStance(self.Owner.Stances[self.World.RenderPlayer ?? self.World.LocalPlayer]))
```
Comment (rob-v) [inline OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs]: This will fail in replays, both RenderPlayer and LocalPlayer can be null.

**Comment by atlimit8** [comment]
Added replay handling.

File: OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs [hunk 3]
```diff
@@ -18,38 +18,43 @@
...
+		protected override void Created(Actor self)
 		{
+			base.Created(self);
 			modifier = self.TraitsImplementing<IRadarColorModifier>().FirstOrDefault();
 		}
 
 		public IEnumerable<Pair<CPos, Color>> RadarSignatureCells(Actor self)
 		{
+			var viewer = self.World.RenderPlayer ?? self.World.LocalPlayer;
+			if (IsTraitDisabled || !(viewer != null && Info.ValidStances.HasStance(self.Owner.Stances[viewer])))
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs]: Using `IsTraitDisabled || viewer == null || !Info.ValidStances...` looks simpler to me (I don't really mind this either, though).

Comment (rob-v) [inline OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs]: It is wrong now, because now nothing will be displayed on radar in replay for 'Disable shroud'.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs]: I believe the convention we have set elsewhere is that 'Disable shrould' is treated like an ally.

Comment (pchote) [inline OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs]: The fix is to move the `!` to the `Info.ValidStances`: (`(IsTraitDisabled || (viewer != null && !Info.ValidStances.HasStance(self.Owner.Stances[viewer])))`)

**Review by pchote — DISMISSED** [review]
Works as advertised, otherwise.

**Comment by atlimit8** [comment]
Applied fix-up.

File: OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs [hunk 4]
```diff
@@ -18,38 +18,43 @@
...
 		IRadarColorModifier modifier;
 
 		Color currentColor = Color.Transparent;
 		Func<Pair<CPos, SubCell>, Pair<CPos, Color>> cellToSignature;
 
-		public AppearsOnRadar(AppearsOnRadarInfo info)
-		{
-			this.info = info;
-		}
+		public AppearsOnRadar(AppearsOnRadarInfo info) : base(info) { }
```
Comment (rob-v) [inline OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs]: small note: base(info) is in most cases on new line

**Review by rob-v — APPROVED** [review]
👍

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f2040025e6e079e9e9f407aa6c627128e546bdda...e86d5c912ca2458585740530e730309f1c39abb4

**Pull request merged**